lib/common/nvpair.c

```diff
@@ -649,42 +649,6 @@ crm_element_value_copy(const xmlNode *data, const char *name)
     return value_copy;
 }
 
-/*!
- * \brief Add hash table entry to XML as (possibly legacy) name/value
- *
- * Suitable for \c g_hash_table_foreach(), this function takes a hash table key
- * and value, with an XML node passed as user data, and adds an XML attribute
- * with the specified name and value if it does not already exist. If the key
- * name starts with a digit, this will instead add a \<param name=NAME
- * value=VALUE/> child to the XML (for legacy compatibility with heartbeat).
- *
- * \param[in]     key        Key of hash table entry
- * \param[in]     value      Value of hash table entry
- * \param[in,out] user_data  XML node
- */
-void
-hash2smartfield(gpointer key, gpointer value, gpointer user_data)
-{
-    const char *name = key;
-    const char *s_value = value;
-
-    xmlNode *xml_node = user_data;
-
-    if (isdigit(name[0])) {
```

Comment:
We need to keep this. It's not actually heartbeat-specific, it's handling any numeric parameter name (OCF doesn't impose any requirements on parameter names). XML attribute names can't start with a number, so this adds any such parameters as <param name=N value=V> children instead of N=V XML attributes. xml2list() then parses it. We should update the comment. (The old heartbeat-class resources -- agents provided with heartbeat, not actually directly related to the cluster layer -- used positional parameters, so this was added for them.)

We're not even doing a good job of it here. XML attribute names must start with a letter or underscore, and can only contain certain characters. Rather than trying to check that, we should probably just use the <param> children always. I *think* that would be backward-compatible, but it's probably not worth the hassle to verify right now.

Reply:
Silly me for believing a comment.

I'm glad you've been working on this long enough to remember why things are a certain way.

Reply:
> I think that would be backward-compatible, but it's probably not worth the hassle to verify right now.

Hard to think of why it *wouldn't* be. But I agree. That'll just be a TODO comment.

Reply:
Also sounds like we should eventually be using `hash2smartfield()` everywhere we're currently using `hash2field()`. That's for later. Would need to make sure there's an `xml2list()` on the other end of all those calls.